File: packages/csharp-sdk/SchlepClient.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using Microsoft.Extensions.Logging;
using SchlepEngine.Exceptions;
using SchlepEngine.Types;

namespace SchlepEngine;

/// <summary>
/// Main client for interacting with the Schlep-engine API.
/// </summary>
/// <remarks>
/// The client provides methods for uploading data, training models, deploying models,
/// checking job status, and streaming real-time events.
///
/// <para><strong>Authentication</strong></para>
/// <para>The client requires an API key for authentication. You can provide it either:</para>
/// <list type="bullet">
/// <item>As a parameter when creating the client: <c>new SchlepClient("your-api-key")</c></item>
/// <item>Via the <c>SCHLEP_API_KEY</c> environment variable</item>
/// </list>
///
/// <para><strong>Example Usage</strong></para>
/// <code>
/// var client = new SchlepClient("your-api-key");
///
/// // Upload data
/// var uploadResult = await client.UploadAsync("Hello, world!");
/// Console.WriteLine($"Upload job ID: {uploadResult.JobId}");
///
/// // Train a model
/// var config = new TrainConfig
/// {
///     ModelType = "classification",
///     DatasetId = uploadResult.JobId
/// };
/// var trainResult = await client.TrainAsync(config);
///
/// // Deploy model
/// if (trainResult.ModelId != null)
/// {
///     var deployResult = await client.DeployAsync(trainResult.ModelId);
///     Console.WriteLine($"Model deployed at: {deployResult.EndpointUrl}");
/// }
/// </code>
/// </remarks>
public class SchlepClient : IDisposable
{
    private const string DefaultBaseUrl = "https://api.schlep-engine.com/v1";

    private readonly HttpClient _httpClient;
    private readonly ILogger<SchlepClient>? _logger;
    private readonly JsonSerializerOptions _jsonOptions;
    private readonly string _baseUrl;
    private readonly string _apiKey;
    private bool _disposed;

    /// <summary>
    /// Initializes a new instance of the <see cref="SchlepClient"/> class with the provided API key.
    /// </summary>
    /// <param name="apiKey">Your Schlep-engine API key.</param>
    /// <param name="httpClient">Optional HTTP client instance. If not provided, a new one will be created.</param>
    /// <param name="logger">Optional logger instance.</param>
    /// <exception cref="ConfigurationException">Thrown if the API key is null or empty.</exception>
    public SchlepClient(string apiKey, HttpClient? httpClient = null, ILogger<SchlepClient>? logger = null)
        : this(apiKey, DefaultBaseUrl, httpClient, logger)
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="SchlepClient"/> class with a custom base URL.
    /// </summary>
    /// <param name="apiKey">Your Schlep-engine API key.</param>
    /// <param name="baseUrl">Custom base URL for the API.</param>
    /// <param name="httpClient">Optional HTTP client instance. If not provided, a new one will be created.</param>
    /// <param name="logger">Optional logger instance.</param>
    /// <exception cref="ConfigurationException">Thrown if the API key is null or empty.</exception>
    public SchlepClient(string apiKey, string baseUrl, HttpClient? httpClient = null, ILogger<SchlepClient>? logger = null)
    {
        if (string.IsNullOrWhiteSpace(apiKey))
            throw new ConfigurationException("API key cannot be null or empty");

        _apiKey = apiKey;
        _baseUrl = baseUrl;
        _logger = logger;

        _httpClient = httpClient ?? new HttpClient();
        _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
        _httpClient.DefaultRequestHeaders.Accept.Clear();
        _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        _jsonOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
            WriteIndented = false
        };

        _logger?.LogDebug("SchlepClient initialized with base URL: {BaseUrl}", _baseUrl);
    }

    /// <summary>
    /// Creates a new client using the API key from the SCHLEP_API_KEY environment variable.
    /// </summary>
    /// <param name="httpClient">Optional HTTP client instance.</param>
    /// <param name="logger">Optional logger instance.</param>
    /// <returns>A new SchlepClient instance.</returns>
    /// <exception cref="ConfigurationException">Thrown if the environment variable is not set or empty.</exception>
    public static SchlepClient FromEnvironment(HttpClient? httpClient = null, ILogger<SchlepClient>? logger = null)
    {
        var apiKey = Environment.GetEnvironmentVariable("SCHLEP_API_KEY");
        if (string.IsNullOrWhiteSpace(apiKey))
            throw new ConfigurationException("SCHLEP_API_KEY environment variable not set or empty");

        return new SchlepClient(apiKey, httpClient, logger);
    }

    /// <summary>
    /// Uploads data to Schlep-engine for processing.
    /// </summary>
    /// <param name="data">The data to upload (can be text, JSON, etc.).</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>UploadResponse containing the job ID and status.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    /// <exception cref="HttpRequestException">Thrown if there's a network error.</exception>
    public async Task<UploadResponse> UploadAsync(string data, CancellationToken cancellationToken = default)
    {
        _logger?.LogDebug("Uploading data");

        var request = new UploadRequest { Data = data };
        var url = $"{_baseUrl}/upload";

        return await PostAsync<UploadRequest, UploadResponse>(url, request, cancellationToken);
    }

    /// <summary>
    /// Trains a machine learning model with the provided configuration.
    /// </summary>
    /// <param name="config">Training configuration (model type, dataset, parameters).</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>TrainResponse containing the job ID and status.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    /// <exception cref="HttpRequestException">Thrown if there's a network error.</exception>
    public async Task<TrainResponse> TrainAsync(TrainConfig config, CancellationToken cancellationToken = default)
    {
        _logger?.LogDebug("Training model with config: {@Config}", config);

        var url = $"{_baseUrl}/train";
        return await PostAsync<TrainConfig, TrainResponse>(url, config, cancellationToken);
    }

    /// <summary>
    /// Trains a machine learning model with JSON configuration.
    /// </summary>
    /// <param name="configJson">JSON configuration as a string.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>TrainResponse containing the job ID and status.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    /// <exception cref="HttpRequestException">Thrown if there's a network error.</exception>
    public async Task<TrainResponse> TrainAsync(string configJson, CancellationToken cancellationToken = default)
    {
        _logger?.LogDebug("Training model with JSON config");

        var url = $"{_baseUrl}/train";
        var content = new StringContent(configJson, Encoding.UTF8, "application/json");

        using var response = await _httpClient.PostAsync(url, content, cancellationToken);
        return await HandleResponseAsync<TrainResponse>(response);
    }

    /// <summary>
    /// Deploys a trained model to a production endpoint.
    /// </summary>
    /// <param name="modelId">ID of the trained model to deploy.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>DeployResponse containing the deployment details.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    /// <exception cref="HttpRequestException">Thrown if there's a network error.</exception>
    public async Task<DeployResponse> DeployAsync(string modelId, CancellationToken cancellationToken = default)
    {
        _logger?.LogDebug("Deploying model: {ModelId}", modelId);

        var request = new DeployRequest { ModelId = modelId };
        var url = $"{_baseUrl}/deploy";

        return await PostAsync<DeployRequest, DeployResponse>(url, request, cancellationToken);
    }

    /// <summary>
    /// Checks the status of a job (upload, training, deployment, etc.).
    /// </summary>
    /// <param name="jobId">ID of the job to check.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>StatusResponse containing the current job status and progress.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    /// <exception cref="HttpRequestException">Thrown if there's a network error.</exception>
    public async Task<StatusResponse> StatusAsync(string jobId, CancellationToken cancellationToken = default)
    {
        _logger?.LogDebug("Checking status for job: {JobId}", jobId);

        var url = $"{_baseUrl}/status/{jobId}";

        using var response = await _httpClient.GetAsync(url, cancellationToken);
        return await HandleResponseAsync<StatusResponse>(response);
    }

    /// <summary>
    /// Streams real-time events from Schlep-engine.
    /// </summary>
    /// <param name="config">Configuration for the types of events to stream.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <remarks>
    /// This is a basic WebSocket streaming implementation. For production use,
    /// you may want to implement more sophisticated event handling and reconnection logic.
    /// </remarks>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    /// <exception cref="HttpRequestException">Thrown if there's a network error.</exception>
    public async Task StreamAsync(StreamConfig config, CancellationToken cancellationToken = default)
    {
        _logger?.LogDebug("Starting stream with config: {@Config}", config);

        // This is a simplified implementation for demonstration
        // In practice, you'd implement proper WebSocket handling here
        var wsUrl = _baseUrl.Replace("https://", "wss://").Replace("http://", "ws://") + "/stream";

        _logger?.LogInformation("WebSocket connection would be established at: {WsUrl}", wsUrl);
        _logger?.LogInformation("Stream configuration: {@Config}", config);

        // TODO: Implement actual WebSocket streaming
        // For now, this is a placeholder implementation
        await Task.CompletedTask;
    }

    /// <summary>
    /// Gets the base URL for this client.
    /// </summary>
    public string BaseUrl => _baseUrl;

    /// <summary>
    /// Performs a POST request with JSON payload.
    /// </summary>
    private async Task<TResponse> PostAsync<TRequest, TResponse>(string url, TRequest request, CancellationToken cancellationToken)
    {
        var json = JsonSerializer.Serialize(request, _jsonOptions);
        var content = new StringContent(json, Encoding.UTF8, "application/json");

        using var response = await _httpClient.PostAsync(url, content, cancellationToken);
        return await HandleResponseAsync<TResponse>(response);
    }

    /// <summary>
    /// Handles HTTP response and parses JSON or throws appropriate exception.
    /// </summary>
    private async Task<T> HandleResponseAsync<T>(HttpResponseMessage response)
    {
        var responseBody = await response.Content.ReadAsStringAsync();

        if (response.IsSuccessStatusCode)
        {
            try
            {
                return JsonSerializer.Deserialize<T>(responseBody, _jsonOptions)
                    ?? throw new ApiException((int)response.StatusCode, "Failed to deserialize response");
            }
            catch (JsonException ex)
            {
                throw new ApiException((int)response.StatusCode, $"Failed to parse response: {ex.Message}");
            }
        }
        else
        {
            // Try to parse error response
            try
            {
                using var doc = JsonDocument.Parse(responseBody);
                var message = doc.RootElement.TryGetProperty("message", out var msgElement)
                    ? msgElement.GetString() ?? "Unknown API error"
                    : "Unknown API error";
                throw new ApiException((int)response.StatusCode, message);
            }
            catch (JsonException)
            {
                throw new ApiException((int)response.StatusCode, responseBody);
            }
        }
    }

    /// <summary>
    /// Releases all resources used by the <see cref="SchlepClient"/>.
    /// </summary>
    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    /// <summary>
    /// Releases the unmanaged resources and optionally releases the managed resources.
    /// </summary>
    /// <param name="disposing">true to release both managed and unmanaged resources; false to release only unmanaged resources.</param>
    protected virtual void Dispose(bool disposing)
    {
        if (!_disposed && disposing)
        {
            _httpClient?.Dispose();
            _logger?.LogDebug("SchlepClient disposed");
            _disposed = true;
        }
    }
}

File: labs/packages/java-sdk/src/main/java/io/schlepengine/SchlepClient.java
package io.schlepengine;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.schlepengine.api.*;
import io.schlepengine.exceptions.ApiException;
import io.schlepengine.exceptions.ConfigurationException;
import io.schlepengine.types.*;
import okhttp3.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

/**
 * Main client for interacting with the Schlep-engine API.
 *
 * <p>The client provides methods for uploading data, training models, deploying models,
 * checking job status, and streaming real-time events.</p>
 *
 * <h3>Authentication</h3>
 * <p>The client requires an API key for authentication. You can provide it either:</p>
 * <ul>
 *   <li>As a parameter when creating the client: {@code new SchlepClient("your-api-key")}</li>
 *   <li>Via the {@code SCHLEP_API_KEY} environment variable</li>
 * </ul>
 *
 * <h3>Example Usage</h3>
 * <pre>{@code
 * SchlepClient client = new SchlepClient("your-api-key");
 *
 * // Upload data
 * UploadResponse uploadResult = client.upload("Hello, world!");
 * System.out.println("Upload job ID: " + uploadResult.getJobId());
 *
 * // Train a model
 * TrainConfig config = new TrainConfig()
 *     .setModelType("classification")
 *     .setDatasetId(uploadResult.getJobId());
 * TrainResponse trainResult = client.train(config);
 *
 * // Deploy model
 * if (trainResult.getModelId() != null) {
 *     DeployResponse deployResult = client.deploy(trainResult.getModelId());
 *     System.out.println("Model deployed at: " + deployResult.getEndpointUrl());
 * }
 * }</pre>
 *
 * @author Schlep-engine Team
 * @version 1.0.0
 * @since 1.0.0
 */
public class SchlepClient {
    private static final Logger logger = LoggerFactory.getLogger(SchlepClient.class);

    public static final String DEFAULT_BASE_URL = "https://api.schlep-engine.com/v1";

    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final String baseUrl;
    private final String apiKey;

    // API clients
    private DataProcessingClient dataProcessingClient;
    private MLPipelineClient mlPipelineClient;
    private AnalyticsClient analyticsClient;
    private DocumentClient documentClient;
    private QualityClient qualityClient;
    private StorageClient storageClient;
    private MonitoringClient monitoringClient;
    private UsersClient usersClient;
    private AdminClient adminClient;

    /**
     * Create a new Schlep-engine client with the provided API key.
     *
     * @param apiKey Your Schlep-engine API key
     * @throws ConfigurationException if the API key is null or empty
     */
    public SchlepClient(String apiKey) throws ConfigurationException {
        this(apiKey, DEFAULT_BASE_URL);
    }

    /**
     * Create a new client using the API key from the SCHLEP_API_KEY environment variable.
     *
     * @return A new SchlepClient instance
     * @throws ConfigurationException if the environment variable is not set or empty
     */
    public static SchlepClient fromEnv() throws ConfigurationException {
        String apiKey = System.getenv("SCHLEP_API_KEY");
        if (apiKey == null || apiKey.trim().isEmpty()) {
            throw new ConfigurationException("SCHLEP_API_KEY environment variable not set or empty");
        }
        return new SchlepClient(apiKey);
    }

    /**
     * Create a new client with a custom base URL.
     *
     * @param apiKey Your Schlep-engine API key
     * @param baseUrl Custom base URL for the API
     * @throws ConfigurationException if the API key is null or empty
     */
    public SchlepClient(String apiKey, String baseUrl) throws ConfigurationException {
        if (apiKey == null || apiKey.trim().isEmpty()) {
            throw new ConfigurationException("API key cannot be null or empty");
        }

        this.apiKey = apiKey;
        this.baseUrl = baseUrl;
        this.objectMapper = new ObjectMapper();
        this.httpClient = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(60, TimeUnit.SECONDS)
            .build();

        logger.debug("SchlepClient initialized with base URL: {}", baseUrl);
    }

    /**
     * Upload data to Schlep-engine for processing.
     *
     * @param data The data to upload (can be text, JSON, etc.)
     * @return UploadResponse containing the job ID and status
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public UploadResponse upload(String data) throws ApiException, IOException {
        logger.debug("Uploading data");

        UploadRequest request = new UploadRequest().setData(data);
        String url = baseUrl + "/upload";

        RequestBody body = RequestBody.create(
            objectMapper.writeValueAsString(request),
            MediaType.get("application/json")
        );

        Request httpRequest = new Request.Builder()
            .url(url)
            .post(body)
            .addHeader("Authorization", "Bearer " + apiKey)
            .addHeader("Content-Type", "application/json")
            .build();

        return executeRequest(httpRequest, UploadResponse.class);
    }

    /**
     * Train a machine learning model with the provided configuration.
     *
     * @param config Training configuration (model type, dataset, parameters)
     * @return TrainResponse containing the job ID and status
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public TrainResponse train(TrainConfig config) throws ApiException, IOException {
        logger.debug("Training model with config: {}", config);

        String url = baseUrl + "/train";

        RequestBody body = RequestBody.create(
            objectMapper.writeValueAsString(config),
            MediaType.get("application/json")
        );

        Request httpRequest = new Request.Builder()
            .url(url)
            .post(body)
            .addHeader("Authorization", "Bearer " + apiKey)
            .addHeader("Content-Type", "application/json")
            .build();

        return executeRequest(httpRequest, TrainResponse.class);
    }

    /**
     * Train a machine learning model with JSON configuration.
     *
     * @param configJson JSON configuration as a string
     * @return TrainResponse containing the job ID and status
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public TrainResponse train(String configJson) throws ApiException, IOException {
        logger.debug("Training model with JSON config");

        String url = baseUrl + "/train";

        RequestBody body = RequestBody.create(
            configJson,
            MediaType.get("application/json")
        );

        Request httpRequest = new Request.Builder()
            .url(url)
            .post(body)
            .addHeader("Authorization", "Bearer " + apiKey)
            .addHeader("Content-Type", "application/json")
            .build();

        return executeRequest(httpRequest, TrainResponse.class);
    }

    /**
     * Deploy a trained model to a production endpoint.
     *
     * @param modelId ID of the trained model to deploy
     * @return DeployResponse containing the deployment details
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public DeployResponse deploy(String modelId) throws ApiException, IOException {
        logger.debug("Deploying model: {}", modelId);

        DeployRequest request = new DeployRequest().setModelId(modelId);
        String url = baseUrl + "/deploy";

        RequestBody body = RequestBody.create(
            objectMapper.writeValueAsString(request),
            MediaType.get("application/json")
        );

        Request httpRequest = new Request.Builder()
            .url(url)
            .post(body)
            .addHeader("Authorization", "Bearer " + apiKey)
            .addHeader("Content-Type", "application/json")
            .build();

        return executeRequest(httpRequest, DeployResponse.class);
    }

    /**
     * Check the status of a job (upload, training, deployment, etc.).
     *
     * @param jobId ID of the job to check
     * @return StatusResponse containing the current job status and progress
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public StatusResponse status(String jobId) throws ApiException, IOException {
        logger.debug("Checking status for job: {}", jobId);

        String url = baseUrl + "/status/" + jobId;

        Request httpRequest = new Request.Builder()
            .url(url)
            .get()
            .addHeader("Authorization", "Bearer " + apiKey)
            .build();

        return executeRequest(httpRequest, StatusResponse.class);
    }

    /**
     * Stream real-time events from Schlep-engine.
     *
     * <p>This is a basic WebSocket streaming implementation. For production use,
     * you may want to implement more sophisticated event handling and reconnection logic.</p>
     *
     * @param config Configuration for the types of events to stream
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public void stream(StreamConfig config) throws ApiException, IOException {
        logger.debug("Starting stream with config: {}", config);

        // This is a simplified implementation for demonstration
        // In practice, you'd implement proper WebSocket handling here
        String wsUrl = baseUrl.replace("https://", "wss://").replace("http://", "ws://") + "/stream";

        logger.info("WebSocket connection would be established at: {}", wsUrl);
        logger.info("Stream configuration: {}", objectMapper.writeValueAsString(config));

        // TODO: Implement actual WebSocket streaming
        // For now, this is a placeholder implementation
    }

    /**
     * Execute an HTTP request and parse the response.
     *
     * @param request The HTTP request to execute
     * @param responseClass The class to deserialize the response into
     * @param <T> The response type
     * @return The parsed response object
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    private <T> T executeRequest(Request request, Class<T> responseClass) throws ApiException, IOException {
        try (Response response = httpClient.newCall(request).execute()) {
            String responseBody = response.body() != null ? response.body().string() : "";

            if (response.isSuccessful()) {
                return objectMapper.readValue(responseBody, responseClass);
            } else {
                // Try to parse error response
                try {
                    JsonNode errorJson = objectMapper.readTree(responseBody);
                    String message = errorJson.has("message") ?
                        errorJson.get("message").asText() : "Unknown API error";
                    throw new ApiException(response.code(), message);
                } catch (Exception e) {
                    throw new ApiException(response.code(), responseBody);
                }
            }
        }
    }

    /**
     * Get the base URL for this client.
     *
     * @return The base URL
     */
    public String getBaseUrl() {
        return baseUrl;
    }

    /**
     * Get the data processing API client.
     *
     * @return DataProcessingClient instance
     */
    public DataProcessingClient data() {
        if (dataProcessingClient == null) {
            dataProcessingClient = new DataProcessingClient(httpClient, objectMapper, baseUrl, apiKey);
        }
        return dataProcessingClient;
    }

    /**
     * Get the ML pipeline API client.
     *
     * @return MLPipelineClient instance
     */
    public MLPipelineClient ml() {
        if (mlPipelineClient == null) {
            mlPipelineClient = new MLPipelineClient(httpClient, objectMapper, baseUrl, apiKey);
        }
        return mlPipelineClient;
    }

    /**
     * Get the analytics API client.
     *
     * @return AnalyticsClient instance
     */
    public AnalyticsClient analytics() {
        if (analyticsClient == null) {
            analyticsClient = new AnalyticsClient(httpClient, objectMapper, baseUrl, apiKey);
        }
        return analyticsClient;
    }

    /**
     * Get the document extraction API client.
     *
     * @return DocumentClient instance
     */
    public DocumentClient document() {
        if (documentClient == null) {
            documentClient = new DocumentClient(httpClient, objectMapper, baseUrl, apiKey);
        }
        return documentClient;
    }

    /**
     * Get the data quality API client.
     *
     * @return QualityClient instance
     */
    public QualityClient quality() {
        if (qualityClient == null) {
            qualityClient = new QualityClient(httpClient, objectMapper, baseUrl, apiKey);
        }
        return qualityClient;
    }

    /**
     * Get the storage API client.
     *
     * @return StorageClient instance
     */
    public StorageClient storage() {
        if (storageClient == null) {
            storageClient = new StorageClient(httpClient, objectMapper, baseUrl, apiKey);
        }
        return storageClient;
    }

    /**
     * Get the monitoring API client.
     *
     * @return MonitoringClient instance
     */
    public MonitoringClient monitoring() {
        if (monitoringClient == null) {
            monitoringClient = new MonitoringClient(httpClient, objectMapper, baseUrl, apiKey);
        }
        return monitoringClient;
    }

    /**
     * Get the users API client.
     *
     * @return UsersClient instance
     */
    public UsersClient users() {
        if (usersClient == null) {
            usersClient = new UsersClient(httpClient, objectMapper, baseUrl, apiKey);
        }
        return usersClient;
    }

    /**
     * Get the admin API client.
     *
     * @return AdminClient instance
     */
    public AdminClient admin() {
        if (adminClient == null) {
            adminClient = new AdminClient(httpClient, objectMapper, baseUrl, apiKey);
        }
        return adminClient;
    }

    /**
     * Close the client and release resources.
     */
    public void close() {
        httpClient.dispatcher().executorService().shutdown();
        httpClient.connectionPool().evictAll();
        logger.debug("SchlepClient closed");
    }
}